File: src/boj/boj_1987.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_1987 {
    static final int[] dr = {1, 0, -1, 0};
    static final int[] dc = {0, 1, 0, -1};

    static int answer = 0;


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int r = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());

        char[][] map = new char[r][];
        for (int cr = 0; cr < r; cr++) {
            map[cr] = br.readLine().toCharArray();
        }

        int[] visited = new int[26];
        visited[map[0][0] - 'A'] = 1;
        move(r, c, map, visited, 0, 0);

        System.out.println(answer);
        br.close();
    }

    static void move(int r, int c, char[][] map, int[] visited, int cr, int cc) {
        answer = Math.max(answer, visited[map[cr][cc] - 'A']);
        for (int d = 0; d < 4; d++) {
            int nr = cr + dr[d];
            int nc = cc + dc[d];
            if (0 <= nr && nr < r && 0 <= nc && nc < c && visited[map[nr][nc] - 'A'] == 0) {
                visited[map[nr][nc] - 'A'] = visited[map[cr][cc] - 'A'] + 1;
                move(r, c, map, visited, nr, nc);
                visited[map[nr][nc] - 'A'] = 0;
            }
        }

    }
}

File: src/boj/boj_1987.kt
package boj

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max
import kotlin.properties.Delegates

private val dr = arrayOf(0, 0, -1, 1)
private val dc = arrayOf(1, -1, 0, 0)
private var r by Delegates.notNull<Int>()
private var c by Delegates.notNull<Int>()
private lateinit var map: Array<String>
private lateinit var visited: BooleanArray
private var answer = 1

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    with(StringTokenizer(readLine())) {
        r = nextToken().toInt()
        c = nextToken().toInt()
    }
    map = Array(r) { readLine() }
    visited = BooleanArray(26).apply {
        this[map[0][0] - 'A'] = true
    }

    bt(1, 0, 0)

    println(answer)
    close()
}

private fun bt(depth: Int, curR: Int, curC: Int) {
    answer = max(answer, depth)
    for (d in 0 until 4) {
        val nr = curR + dr[d]
        val nc = curC + dc[d]
        if (nr in 0 until r && nc in 0 until c) {
            val visitedIndex = map[nr][nc] - 'A'
            if (!visited[visitedIndex]) {
                visited[visitedIndex] = true
                bt(depth + 1, nr, nc)
                visited[visitedIndex] = false
            }
        }
    }
}
